### PR TITLE
Prototype: target typed "default"

### PR DESCRIPTION
**Pull request by jcouv**
Allow `int x = default;` and `M(default);` where the type of `default` is inferred.

Design notes (updated):
- Introduces a new syntax node for target-typed default.
- During initial binding, this is represented by a `BoundDefaultOperator` without a type.
- A new kind of conversion is introduced (`DefaultLiteral`). "default" can convert to any struct and non-nullable type.
- During lowering, this conversion produces a `BoundDefaultOperator` with the target type of the conversion.
- "default" contributes to overload resolution, by excluding parameters that are nullable.
- Some language question recorded in https://github.com/dotnet/roslyn/issues/13602

CC @dotnet/roslyn-compiler for review. This is low-pri (not for C# 7).

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 1]
```diff
@@ -918,6 +918,21 @@
       <summary>Creates an DefaultExpressionSyntax node.</summary>
     </FactoryComment>
   </Node>
+  <Node Name="TargetTypedDefaultExpressionSyntax" Base="ExpressionSyntax">
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: >TargetTypedDefaultExpressionSyntax [](start = 14, length = 34)

I suggest a more likely name in the grammar is "default literal".

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 2]
```diff
@@ -97,6 +98,11 @@ internal partial class Binder
                 return CreateTupleLiteralConversion(syntax, (BoundTupleLiteral)source, conversion, isCast, destination, diagnostics);
             }
 
+            if (conversion.Kind == ConversionKind.ImplicitDefault)
+            {
+                return new BoundDefaultOperator(source.Syntax, destination);
+            }
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs]: You appear to be throwing away the default operator and the conversion, and replacing it with its lowering here. The design pattern is to do binding only in the binder, and lowering in the local lowering phase. This pattern will make it difficult to get the correct result in the semantic model.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs]: I think `default` should be handled similarly to the way `null` is handled.

---
In reply to: [77467548](https://github.com/dotnet/roslyn/pull/13603#discussion_r77467548) [](ancestors = 77467548)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 3]
```diff
@@ -2247,6 +2256,11 @@ private BoundExpression BindArgumentExpression(DiagnosticBag diagnostics, Expres
 
                     arguments[arg] = ((OutVarLocalPendingInference)argument).SetInferredType(parameterType, success: !hasErrors);
                 }
+                else if (argument.Kind == BoundKind.DefaultPendingInference)
+                {
+                    TypeSymbol parameterType = GetCorrespondingParameterType(ref result, parameters, arg);
+                    arguments[arg] = ((BoundDefaultPendingInference)argument).SetInferredType(parameterType);
+                }
                 else if (argument.Kind == BoundKind.OutDeconstructVarPendingInference)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: This is the wrong place to handle this. It should be handled as a "conversion from expression", like `null`.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: ... because it should not be specified to work with only specific language contexts.

---
In reply to: [77467707](https://github.com/dotnet/roslyn/pull/13603#discussion_r77467707) [](ancestors = 77467707)

**Comment by gafter** [comment]
> 

These cases should be unified. The problem is that the left-hand-side doesn't have a type. Use BoundExpression.Display for the string form of the problematic expression.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4864 in 14592f6. [](commit_id = 14592f690d0ac89b11838fc37a68b4c40cf07da0, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 4]
```diff
@@ -114,7 +114,7 @@ private static ImmutableArray<MethodSymbol> GetOriginalMethods(OverloadResolutio
             boundExpression.WasCompilerGenerated = true;
 
             var analyzedArguments = AnalyzedArguments.GetInstance();
-            Debug.Assert(!args.Any(e => e.Kind == BoundKind.OutVarLocalPendingInference || e.Kind == BoundKind.OutDeconstructVarPendingInference));
+            Debug.Assert(!args.Any(e => e.Kind == BoundKind.OutVarLocalPendingInference || e.Kind == BoundKind.OutDeconstructVarPendingInference || e.Kind == BoundKind.DefaultPendingInference));
             analyzedArguments.Arguments.AddRange(args);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: In a bad invocation, it should be capable of remaining in the bound tree.

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 5]
```diff
@@ -348,6 +348,10 @@ private ImmutableArray<BoundExpression> BuildArgumentsForDynamicInvocation(Analy
                         {
                             builder[i] = ((OutVarLocalPendingInference)argument).FailInference(this, diagnostics);
                         }
+                        if (argument.Kind == BoundKind.DefaultPendingInference)
+                        {
+                            builder[i] = ((BoundDefaultPendingInference)argument).FailInference(this, diagnostics);
+                        }
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: Here, too, the problem is an expression without a type, not this particular expression form.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 6]
```diff
@@ -2738,6 +2738,16 @@ internal BoundExpression GenerateConversionForAssignment(TypeSymbol targetType,
                 }
             }
 
+            if (operand.Kind == BoundKind.DefaultPendingInference)
+            {
+                // PROTOTYPE(default) Need to handle type parameters here (`where T : struct` for instance)
+                if (!targetType.IsStructType() || targetType.IsNullableType())
+                {
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: I think the error only arises when the type is nullable.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 7]
```diff
@@ -1229,7 +1229,7 @@ private static TypeSymbol GetParameterType(int argIndex, MemberAnalysisResult re
                 bool okToDowngradeToNeither;
                 BetterResult r;
 
-                if (argumentKind == BoundKind.OutVarLocalPendingInference || argumentKind == BoundKind.OutDeconstructVarPendingInference)
+                if (argumentKind == BoundKind.OutVarLocalPendingInference || argumentKind == BoundKind.OutDeconstructVarPendingInference || argumentKind == BoundKind.DefaultPendingInference)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: >argumentKind == BoundKind [](start = 140, length = 25)

I believe this should be handled in BetterConverionFromExpression.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 8]
```diff
@@ -2980,6 +2980,18 @@ private RefKind GetEffectiveParameterRefKind(ParameterSymbol parameter, RefKind
...
+                if (parameterType.IsStructType() && !parameterType.IsNullableType())
+                {
+                    return Conversion.Identity;
+                }
+                else
+                {
+                    return Conversion.NoConversion;
+                }
+            }
+
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: This should not be necessary; it should be a consequence of ClassifyImplicitConversionFromExpression, below.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs [hunk 9]
```diff
@@ -943,6 +944,15 @@ private static bool HadLambdaConversionError(DiagnosticBag diagnostics, Analyzed
                         UnwrapIfParamsArray(parameter));
                 }
             }
+            else if (!argument.HasExpressionType() && argument.Kind == BoundKind.DefaultPendingInference)
+            {
+                diagnostics.Add(
+                     ErrorCode.ERR_DefaultIsBadArg,
+                     sourceLocation,
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs]: I think if you revert the changes in this method the resulting diagnostics are of the same quality as for other expressions. In other words, `default` should be handled the same way as `null` for the purposes of this method.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 10]
```diff
@@ -555,6 +555,11 @@
     <Field Name="ConstantValueOpt" Type="ConstantValue" Null="allow"/>
   </Node>
 
+  <Node Name="BoundDefaultPendingInference" Base="BoundExpression">
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: >BoundDefaultPendingInference [](start = 14, length = 28)

You should be able to use the same node, whether inference has been done (i.e. it has a non-null type) or not. In other words, please handle similar to a null expression.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 11]
```diff
@@ -952,6 +953,37 @@ public override void Accept(OperationVisitor visitor)
...
+        internal BoundDefaultOperator FailInference(Binder binder, DiagnosticBag diagnosticsOpt)
+        {
+            if (diagnosticsOpt != null)
+            {
+                Binder.Error(diagnosticsOpt, ErrorCode.ERR_TypeInferenceFailedForImplicitlyTypedDefault, this.Syntax);
+            }
+
+            return this.SetInferredType(binder.CreateErrorType("var"), success: false);
+        }
+    }
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I think this should be done by the machinery around conversion-from-expression.

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs [hunk 12]
```diff
@@ -178,6 +181,9 @@ internal static LanguageVersion RequiredVersion(this MessageID feature)
             // Checks are in the LanguageParser unless otherwise noted.
             switch (feature)
             {
+                case MessageID.IDS_FeatureTargetTypedDefault:
+                    return LanguageVersion.CSharp7;
+
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs]: We usually put it in *either* a feature flag or a language version (not both).

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+        System.Console.Write(x[default]);
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions);
+            comp.VerifyDiagnostics(
+                // (7,32): error CS8301: Cannot infer the type of implicitly-typed default.
+                //         System.Console.Write(x[default]);
+                Diagnostic(ErrorCode.ERR_TypeInferenceFailedForImplicitlyTypedDefault, "default").WithLocation(7, 32)
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: I'm not sure, but I think this should work.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "0");
+        }
+    }
+}
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Other cases to test (that should work)

`switch (2) { case default: ; }`
`if (2 != default) ...`
`var a = new[] { 0, default };`
`void M(int x = default);`
`var a = new int[default];`

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+        public void DynamicInvocation()
+        {
+            string source = @"
+class C
+{
+    static void M1()
+    {
+        dynamic d = null;
+        d.M2(default);
+    }
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: I think we may need to forbid this (until and unless we build runtime support for it). It is like passing a method group.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml (hunk 1 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: That sounds better ;-) I was just hesitant to use the term literal.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs (hunk 15 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Yes. Actually, I forgot to mark this test as skipped for now since it crashes.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 16]
```diff
@@ -881,6 +889,19 @@ private static Conversion ClassifyNullLiteralConversion(BoundExpression source,
             return Conversion.NoConversion;
         }
 
+        private Conversion ClassifyDefaultConversion(BoundDefaultPendingInference source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            Debug.Assert((object)source != null);
+            Debug.Assert((object)destination != null);
+
+            if (destination.IsStructType() && !destination.IsNullableType())
+            {
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: In (at least one of) the corresponding language proposal, there is no such restriction. Not sure if you are proposing to add this as a restriction? Probably requires LDM discussion.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 6 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: I didn't understand this comment. 
The handling of null above checks for `IsValueType`. So here I need to check for `IsStructType`.
`IsNullableType()` only checks if the type is `System.Nullable`

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 11 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Removed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 2 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs]: Makes sense. Initially, I thought that the `default` itself should be updated to carry the type, but now the conversion carries the type that is inferred.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 3 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Removed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs (hunk 5 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: Removed

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs (hunk 7 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: Yes, relying on conversion-from-expression worked. Thanks!

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs (hunk 8 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: Done

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 10 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: Re-used `BoundDefaultOperator`.

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs (hunk 12 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs]: Removed here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs (hunk 14 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Thanks for the suggestions! Added tests.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs (hunk 16 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: From our last discussion, I thought it may be good to avoid ambiguity between `default` and `null`.
The thread on github shows interest to allow `default` anywhere `default(type)` was allowed.
I listed open questions for LDM in #13602

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs (hunk 9 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs]: Reverted. Much simpler now.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs (hunk 13 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Works now

**Comment by jcouv** [comment]
Re-open to trigger CI

**Comment by jcouv** [comment]
@dotnet-bot test this please

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 17]
```diff
@@ -5,6 +5,7 @@
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Roslyn.Utilities;
 using System.Collections.Immutable;
+using System;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs]: You can probably revert this line.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 18]
```diff
@@ -4830,6 +4835,15 @@ private bool IsUsingAliasInScope(string name)
                 return BadExpression(node, boundLeft);
             }
 
+            // PROTOTYPE(default) unify with case above
+            // No member accesses on default
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: I would expect this case (and the following one handling UnboundLambda) to be handled in the call to `MakeMemberAccessValue`. That call should never return a typeless expression.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 6 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Do you intend that default() won't work for enum types?

---
In reply to: [77566856](https://github.com/dotnet/roslyn/pull/13603#discussion_r77566856) [](ancestors = 77566856)

File: src/Compilers/CSharp/Portable/BoundTree/Formatting.cs [hunk 19]
```diff
@@ -119,4 +119,12 @@ public override object Display
             get { throw ExceptionUtilities.Unreachable; }
         }
     }
+
+    internal partial class BoundDefaultOperator
+    {
+        public override object Display
+        {
+            get { return "default"; }
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/Formatting.cs]: >return "default" [](start = 18, length = 16)

Should be `this.Type ?? "default"` (probably with a cast to `object` on one side or the other).

**Comment by gafter** [comment]
:shipit:

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 18 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: I looked into this and it was thorny.
What I tried was to have `MakeMemberAccessValue` return a `BadExpression` and report an error for those two cases. But it made the diagnostics significantly worse in a number of cases (especially lambda). I'll stop by to discuss tomorrow.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 17 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs]: Done

File: src/Compilers/CSharp/Portable/BoundTree/Formatting.cs (hunk 19 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/Formatting.cs]: Fixed. Thanks

**Comment by jcouv** [comment]
@AlekseyTs @cston If you had some time, would you mind doing a second code review. This is going to features/default branch.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 20]
```diff
@@ -2738,6 +2738,22 @@ internal BoundExpression GenerateConversionForAssignment(TypeSymbol targetType,
...
+            if (operand.IsLiteralDefault())
+            {
+                // PROTOTYPE(default) Need to handle type parameters here (`where T : struct` for instance)
+                //if (targetType.TypeKind == TypeKind.TypeParameter)
+                //{
+                //    Error(diagnostics, ErrorCode.ERR_TypeVarCantBeNull, syntax, targetType);
+                //    return;
+                //}
+
+                if (!targetType.IsStructType() || targetType.IsNullableType())
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Would `targetType.IsReferenceType` cover both type parameters and actual types?

Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Would `targetType.IsReferenceType` cover both type parameters and actual types?

Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Would `targetType.IsReferenceType` cover both type parameters and actual types?

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 21]
```diff
@@ -881,6 +893,20 @@ private static Conversion ClassifyNullLiteralConversion(BoundExpression source,
...
+            Debug.Assert((object)source != null);
+            Debug.Assert((object)source.Type == null);
+            Debug.Assert((object)destination != null);
+
+            if (destination.IsStructType() && !destination.IsNullableType())
+            {
+                return Conversion.DefaultLiteral;
+            }
+
+            return Conversion.NoConversion;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: Why are valid conversions limited to struct types?

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs [hunk 22]
```diff
@@ -14,6 +14,11 @@ public static bool IsLiteralNull(this BoundExpression node)
             return node.Kind == BoundKind.Literal && node.ConstantValue.Discriminator == ConstantValueTypeDiscriminator.Null;
         }
 
+        public static bool IsLiteralDefault(this BoundExpression node)
+        {
+            return node.Kind == BoundKind.DefaultOperator && (object)node.Type == null;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs]: Will `node.Type == null` in error situations such as `default(System)`? Should this method check syntax instead?

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 23]
```diff
@@ -444,6 +447,9 @@
   <data name="ERR_ValueCantBeNull" xml:space="preserve">
     <value>Cannot convert null to '{0}' because it is a non-nullable value type</value>
   </data>
+  <data name="ERR_RefCantBeDefault" xml:space="preserve">
+    <value>Cannot convert default to '{0}' because it is a value or nullable type</value>
+  </data>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: `ERR_Ref...` is confusing since it refers to a value type. And why not support `default` for value types?

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs [hunk 24]
```diff
@@ -0,0 +1,604 @@
...
+    static void Main()
+    {
+        int x = (int)default;
+        System.Console.Write(x);
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyEmitDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "0");
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: What is the type of `default` in `int x = (short)default;`?

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 20 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Why not support `default` for reference types?

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Whether `default` should be allowed on reference types is listed as an open LDM question. But the more I discuss with folks and the clearer it is that folks want to allow it. I'll remove this check.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs (hunk 21 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: The idea was that you should use `null` for ref types and `default` for struct types. But there is also an argument that `default` should be allowed anywhere `default(type)` was previous allowed.
I'll remove

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs (hunk 24 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Good question. I expect it should be `short`. I'll put a PROTOTYPE test to remind to check that once semantic model work is done.

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs (hunk 22 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/BoundExpressionExtensions.cs]: I tried that and found that `node.Type` will be `ErrorType` in that scenario.

**Comment by jcouv** [comment]
@gafter @cston, I removed the constraint that `default` should only work on struct types.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 25]
```diff
@@ -4830,6 +4835,15 @@ private bool IsUsingAliasInScope(string name)
...
+            if (boundLeft.IsLiteralDefault())
+            {
+                DiagnosticInfo diagnosticInfo = new CSDiagnosticInfo(ErrorCode.ERR_BadUnaryOp, SyntaxFacts.GetText(operatorToken.Kind()), "default");
+                diagnostics.Add(new CSDiagnostic(diagnosticInfo, operatorToken.GetLocation()));
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Fixed

**Comment by jcouv** [comment]
@cston Would you have time to finish the review? The review feedback so far was addressed :-)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 26]
```diff
@@ -4825,7 +4830,15 @@ private bool IsUsingAliasInScope(string name)
...
+                diagnostics.Add(ErrorCode.ERR_BadUnaryOp, operatorToken.GetLocation(), SyntaxFacts.GetText(operatorToken.Kind()), leftType);
+                return BadExpression(node, boundLeft);
+            }
+
+            // PROTOTYPE(default) unify with case above
+            // No member accesses on default
+            if (boundLeft.IsLiteralDefault())
+            {
+                DiagnosticInfo diagnosticInfo = new CSDiagnosticInfo(ErrorCode.ERR_BadUnaryOp, SyntaxFacts.GetText(operatorToken.Kind()), "default");
                 diagnostics.Add(new CSDiagnostic(diagnosticInfo, operatorToken.GetLocation()));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: `diagnostics.Add(ErrorCode.ERR_BadUnaryOp, ...);`

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 27]
```diff
@@ -918,6 +918,21 @@
...
+      <Kind Name="DefaultKeyword"/>
+      <PropertyComment>
+        <summary>SyntaxToken representing the DefaultKeyword.</summary>
+      </PropertyComment>
+    </Field>
+    <TypeComment>
+      <summary>Class which represents the syntax node for a default literal.</summary>
+    </TypeComment>
+    <FactoryComment>
+      <summary>Creates an DefaultLiteralSyntax node.</summary>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: `Creates a ...`

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs [hunk 28]
```diff
@@ -0,0 +1,613 @@
...
+        return default;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReturnRefType()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Minor point: `int?` is not a reference type.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs [hunk 29]
```diff
@@ -0,0 +1,613 @@
...
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics(
+                // (6,16): error CS0023: Operator '.' cannot be applied to operand of type 'default'
+                //         default.ToString();
+                Diagnostic(ErrorCode.ERR_BadUnaryOp, ".").WithArguments(".", "default").WithLocation(6, 16)
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Perhaps test: `default[0]', `nameof(default)`, `throw default;`

Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Perhaps test: `default[0]', `nameof(default)`, `throw default;`

Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Perhaps test: `default[0]`, `nameof(default)`, `throw default;`

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml (hunk 27 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: Fixed this one and others

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs (hunk 28 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Renamed

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs (hunk 29 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: I always love test ideas :-) Added!

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 26 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Fixed. Thanks

**Comment by jcouv** [comment]
@gafter This is the first part of the prototype on "default", which you had reviewed. The second part is https://github.com/dotnet/roslyn/pull/16425

**Comment by gafter** [comment]
Oh yeah
